src/uct/tcp/tcp_ep.c

```diff
@@ -378,68 +402,132 @@ unsigned uct_tcp_ep_progress_tx(uct_tcp_ep_t *ep)
+        ucs_assert(hdr->length <= (iface->am_buf_size - sizeof(uct_tcp_am_hdr_t)));
 
         if (remainder < sizeof(*hdr) + hdr->length) {
-            break;
+            goto out;
         }
 
         /* Full message was received */
         ep->rx.offset += sizeof(*hdr) + hdr->length;
 
-        if (hdr->am_id >= UCT_AM_ID_MAX) {
-            ucs_error("invalid am id: %d", hdr->am_id);
-            continue;
+        uct_tcp_ep_comp_recv_am(iface, ep, hdr);
+    }
+
+    uct_tcp_ep_ctx_cleanup(&ep->rx);
+
+out:
+    return recv_length > 0;
+}
+
+static inline unsigned uct_tcp_ep_do_partial_rx(uct_tcp_iface_t *iface,
+                                                uct_tcp_ep_t *ep)
+{
+    size_t cur_recvd_length;
+    uct_tcp_am_hdr_t *hdr;
+    size_t recv_length;
+
+    cur_recvd_length = ep->rx.length - ep->rx.offset;
+
+    if (cur_recvd_length < sizeof(*hdr)) {
+        recv_length = sizeof(*hdr) - cur_recvd_length;
+
+        if (!uct_tcp_ep_recv(ep, &recv_length)) {
+            return 0;
+        }
+
+        if (ep->rx.length - ep->rx.offset < sizeof(*hdr)) {
+            return recv_length > 0;
         }
 
-        uct_iface_trace_am(&iface->super, UCT_AM_TRACE_TYPE_RECV, hdr->am_id,
-                           hdr + 1, hdr->length, "RECV fd %d", ep->fd);
-        uct_iface_invoke_am(&iface->super, hdr->am_id, hdr + 1,
-                            hdr->length, 0);
+        hdr = ep->rx.buf + ep->rx.offset;
+
+        if (hdr->length == 0) {
+            goto done;
+        }
+    } else {
+        hdr = ep->rx.buf + ep->rx.offset;
     }
 
-    /* Move the remaining data to the beginning of the buffer
-     * TODO avoid extra copy on partial receive
-     */
-    ucs_assert(remainder >= 0);
-    memmove(ep->rx.buf, ep->rx.buf + ep->rx.offset, remainder);
-    ep->rx.offset = 0;
-    ep->rx.length = remainder;
+    ucs_assert(hdr->length <= (iface->am_buf_size - sizeof(uct_tcp_am_hdr_t)));
+
+    cur_recvd_length = ep->rx.length - ep->rx.offset - sizeof(*hdr);
+    recv_length      = hdr->length - cur_recvd_length;
+
+    if (!uct_tcp_ep_recv(ep, &recv_length)) {
+        return 0;
+    }
+
+    ucs_trace_data("tcp_ep %p: recvd %zu bytes", ep, recv_length);
+
+    if (recv_length == hdr->length - cur_recvd_length) {
+        goto done;
+    }
 
     return recv_length > 0;
+
+done:
+    uct_tcp_ep_comp_recv_am(iface, ep, hdr);
+    uct_tcp_ep_ctx_cleanup(&ep->rx);
+    return 1;
 }
 
+unsigned uct_tcp_ep_progress_rx(uct_tcp_ep_t *ep)
+{
+    uct_tcp_iface_t *iface = ucs_derived_of(ep->super.super.iface,
+                                            uct_tcp_iface_t);
+
+    ucs_trace_func("ep=%p", ep);
+
+    if (uct_tcp_ep_ctx_buf_empty(&ep->rx)) {
+        /* Receive next chunk of data */
+        return uct_tcp_ep_do_next_rx(iface, ep);
+    } else {
+        /* Receive remaining part of AM data */
+        return uct_tcp_ep_do_partial_rx(iface, ep);
```

Comment:
looks like uct_tcp_ep_do_partial_rx() receives only one message, but uct_tcp_ep_do_next_rx() can receive multiple messages?

Reply:
@yosefe @brminich I've updated PR to generalize `uct_tcp_progress_rx` function to receive multiple messages (normal case) and remaining part of the one message (partial receive case) as we discussed with @yosefe offline.
i.e. `uct_tcp_progress_rx` replaces `uct_tcp_do_next_rx` and `uct_tcp_do_partial_rx` .
could you take a look plz?